### ISSUE TITLE
Is it possible to deserialize partial data in its raw format?

### ISSUE THREAD

**Issue by redknightlois** [Microsoft/bond#100]
I know this is a bit unorthodox, but I am not considering Bond as a message passing serialization layer. The raw performance of Bond make it a very interesting serialization technology for other uses too. :+1: on that. 

Lets say I have this object T
```csharp
        public sealed class Documents
        {
            public long Etag;
            public string Key;
            public string Collection;
            public IBonded<string> Data;
        }
```
Now over this think I have to define an index over Key and Collection therefore I need to retrieve the "raw bytes" for Key and Collection (and not the object T) to build an indexing key. 

To give a bit more context, what I have is something like this:
```csharp
_docsSchema = new TableSchema<Documents>("docs")                
            
    .DefineIndex("By/Etag", x => x.Etag)
    .DefineIndex("By/Etag&Collection", x => x.Collection, x => x.Etag)
    .DefineKey(x => x.Key);
```

I dont really care about the types for x.Collection and/or x.Etag or whatever, I do care though about the binary representation of those. And by the way T can be different and so do the indexes defined over them. 

Is it viable to be done using the public API? 
If not possible, How could I accomplish that with the internal API?

**Comment by sapek** [comment]
In general you can get a serialized representation of a value as following.

Given a protocol reader positioned at the beginning of a value:

1) clone the reader
2) call the protocol to skip the value
3) calculate skipped offset
4) call `ReadBytes` on the cloned reader

This is more or less how Bond handles `bonded<T>` although C# version currently doesn't do steps 3 and 4. There is no official API to do 3 but you can use this trick:
```
var clone = reader.Clone();
var before = reader.ReadBytes(0);
reader.Skip(BondDataType.BT_STRING);
var after = reader.ReadBytes(0);
var blob = clone.ReadBytes(after.offset - before.offset);
```
This is not guaranteed to work by the semantics of `IInputStream` but it will work with `InputBuffer` as it is currently implemented. Alternatively, if you have access to the `IInputStream` that the reader is using, then you can just get `Position` property.

At the higher level you would write a custom transform which generates LINQ expressions doing the above to extract the values for the index from the records. There is no documentation/examples for custom transforms yet (on my TODO list) but you can look at the [`IParser`](https://github.com/Microsoft/bond/blob/3c2a105ff49658e5e32abcd16f9026bfeabdcd72/cs/src/core/expressions/IParser.cs#L118) and [`ITransform`](https://github.com/Microsoft/bond/blob/3c2a105ff49658e5e32abcd16f9026bfeabdcd72/cs/src/core/expressions/IParser.cs#L70) interfaces and at [`DeserializerTransform`](https://github.com/Microsoft/bond/blob/master/cs/src/core/expressions/DeserializerTransform.cs) and [`SerializerTransform`](https://github.com/Microsoft/bond/blob/master/cs/src/core/expressions/SerializerTransform.cs) as examples.

If the index definition is static per type you may want to consider using Bond attributes to annotate your schema fields, e.g. (using multi-line string syntax for readability with multiple indices): 
```
struct Documents
{
    [Indices("\
        \ By/Etag \
        \ By/Etag&Collection")]
    0: string Etag;

    [Indices("\
        \ By/Etag&Collection")]
    1: string Collection;  
}
```
Transforms get access to attributes so you wouldn't need to pass that information in your code. 

Doing all this in a transform would let you handle all the decisions based on field names/attributes at the time expressions are generated making the final code you execute on records very fast.

**Comment by redknightlois** [comment]
I have been looking into ```ITrasform``` but I cannot find a way to apply it. [1]

The closest thing to what I have seen in the C++ code is something like this:

```csharp
static void Main()
{
    var obj1 = new Struct10 
    { 
        N = 0x1000, 
        Name = "First Last",
        Items = new List<double> { 3.14, 0 }
    };


    // Serialize an object of old type.
    var output = new OutputBuffer();
    var writer = new CompactBinaryWriter<OutputBuffer>(output);

    Serialize.To<CompactBinaryWriter<OutputBuffer>, Struct10>(writer, obj1);

    // Construct protocol reader and bonded<Struct_2_0> using the buffer
    // containing the serialized old object.
    var input = new InputBuffer(output.Data);
    var reader = new CompactBinaryReader<InputBuffer>(input);

    var obj2 = Deserialize<Struct20>
                    .Apply<ConvertTo>()
                    .From<CompactBinaryReader<InputBuffer>>(reader);
}
```

... but not such construct is available. Where ```ConvertTo``` looks like this

```csharp
public class ConvertTo : ITransform
{
    readonly ITransform parent;

    public ConvertTo(ITransform parent)
    {
        this.parent = parent;
    }

    public Expression Base(IParser parser)
    {
        return parent.Base(parser);
    }

    public Expression Begin
    {
        get { return parent.Begin; }
    }

    public Expression End
    {
        get { return parent.End; }
    }

    public IEnumerable<IField> Fields
    {
        get { return parent.Fields; }
    }

    public Expression UnknownEnd
    {
        get { return parent.UnknownEnd; }
    }

    public Expression UnknownField(IParser parser, Expression fieldType, Expression fieldId)
    {
        return Expression.IfThen( 
                    Expression.Equal( fieldId, Expression.Constant(20) ),
                    Expression.Block(                     
                        // Whatever it should go in here. 
                    )
                );
    }
}
```

Also most of the foundational classes like ```Transform```, ```DeserializerTransform```, etc are a non public. 

What I am missing?


[1] Or write the proper Expression code for that matters. Any "trick" to give a hint on how to convert code into it's Expression constructs.

**Comment by sapek** [comment]
C++ and C# transform are similar at the conceptual level but they are implemented using different techniques. In C# there is much more clear distinction between meta-programming involving transforms and parsers (generating expression trees) and runtime invocation of the delegates created by compiling the expressions.

In particular C# transforms are [applied](https://github.com/Microsoft/bond/blob/3c2a105ff49658e5e32abcd16f9026bfeabdcd72/cs/src/core/expressions/SerializerTransform.cs#L167) at expression-generation-time. The expressions are then [compiled](https://github.com/Microsoft/bond/blob/3c2a105ff49658e5e32abcd16f9026bfeabdcd72/cs/src/core/Serializer.cs#L71-L75) and resulting delegate is [invoked](https://github.com/Microsoft/bond/blob/3c2a105ff49658e5e32abcd16f9026bfeabdcd72/cs/src/core/Serializer.cs#L88) on the schema instance (payload/object).

The built-in transform like `DeserializerTransfrom` and `SerializerTransform` are not meant to be used as base classes. The internal class `Transform` is an implementation detail. From the public API perspective a transform is a class implementing `ITransform` interface. The built-in transforms implement it in a somewhat unorthodox way: the state associated with `ITransform` instance is held in lambda captures. That's what the `Transform` class is used for. That makes for less verbose code albeit possibly harder to follow at first. But you can definitely implement `ITransform` in a traditional way.

**Comment by redknightlois** [comment]
For performance I would like eventually want to have those transformations to happen at expression-generation-time too. What I do not find is how to "Apply" them at all. How do I tell the Serializer that it must use a particular ```ITransform```?

**Comment by sapek** [comment]
Oh, you don't. Not in your case. You would create two classes:

- `Indexer` which would be similar to `Serializer`/`Deserializer`/`Cloner` in that it would hold reference to compiled delegate and expose some method to call it.
- `IndexerTransfrom` which would be used by `Indexer` to create the expression tree to be compiled into the delegate.

So your program would look something like this (pseudo-code):

```
// This is when the IndexTransfrom runs creating expression to index records of the specified type.
var indexer = new Indexer<Documents>();

foreach (record in database)
    // This calls the delegate compiled from the expression created by the transfrom
    indexer.AddToIndex(record);
```

There is a concept in Bond of [custom *serializer* transform](https://microsoft.github.io/bond/manual/bond_cs.html#serializerattribute). This allows the top level `Serialize.To` API to use a different transforms to generate serialization expressions for different protocols. But this doesn't apply to your case because you are not writing custom protocol that needs special handling nor is your operation serialization at all.

**Comment by redknightlois** [comment]
I am 100% sure you are seeing something that I am not. 

Let's see if I get what you are telling me.

When I serialize object ```O``` of type ```T``` using protocol ```P``` I'll get a binary stream ```S={s0,s1,...,sn}``` with ```sk = sizeof(byte)```
Now given that my key ```K``` is composed of ```A``` and ```B``` properties of ```O``` the binary substream should be ```K = { sa,...,sa+sizeof(A), sb, ..., sb + sizeof(B) }```
What I will need to do with the ```Indexer<Documents>``` is actually provide an interface like:

```csharp
public interface IIndexer<T>
{
   R CreateKey<R>( T value );
   R CreateKey<R> ( FastBinaryReader<T> reader ); 
}
```

In effect this implementation is sort-of a mix between the ```Deserializer<T>``` and the ```Cloner<T>``` whereas the ```Cloner<T>``` input gets the type ```T``` and maps into ```T'```, now ```Indexer<T>``` must behave like ```Deserializer<T>``` but only extract those properties that matches the id's on ```T'```. 

Is that right?

**Comment by sapek** [comment]
Here an example of a transform extracting serialized values of fields (it is modified `serializer` example so you can copy-paste into that project to run it):

https://gist.github.com/sapek/3c280db1bfac943bfe56

It is *very* over-simplified but run-able so hopefully it will help you start paying with transforms.

**Comment by redknightlois** [comment]
I actually figured out on Friday that I could (ab)use the current schema migration rules to achieve sort-of what I needed and was going to write the Indexer<T> as an expression that will scan the schema and deal with the known type for key generation (instead of retrieving the data directly, deserialize a simplier schema and then generate it). I will benchmark both approaches during the next week, and let you know. Thanks for the help.

**Comment by sapek** [comment]
Yes, if you are fine deserializing the key that is a perfectly fine approach. [Views](https://microsoft.github.io/bond/manual/compiler.html#struct-views) might come useful.

**Comment by redknightlois** [comment]
I advanced your example and I am almost there, the only issue I have is that I dont know how to retrieve the payload as a byte[].

https://gist.github.com/redknightlois/11d869438015efb9cbc3 

As you can see my Generate should be able to take an array or an OutputBuffer where I must write the payload extracted. However, I cannot find the place to include that. Any idea?

**Comment by sapek** [comment]
You should generate a lambda expression taking two parameters, similar to [`SerializerTransform`](https://github.com/Microsoft/bond/blob/30271563cf84ce4e6e98267c5c916cac8f26ff29/cs/src/core/expressions/SerializerTransform.cs#L76-L79). In your case the second parameter could be `OutputBuffer`.

**Comment by redknightlois** [comment]
I know that it doesnt work (because I tried :D) but shouldnt transcoding over Views work?

Lets say:
```
struct Document
{
    0: int64 Key;
	1: int64 Etag;
	2: string Collection;
    10: string Data;
}

struct DocumentKey view_of Document
{
    Key, Etag;
}
```

with this code:
```csharp
var src = new Document
{
    Key = 999,
    Etag = 10,
    Collection = "Users",
    Data = "{ Name: Tyler, LastName: Durden }"
};

var output = new OutputBuffer();
var writer = new CompactBinaryWriter<OutputBuffer>(output);

Serialize.To(writer, src);

var input = new InputBuffer(output.Data);
var reader = new CompactBinaryReader<InputBuffer>(input);

output = new OutputBuffer();
writer = new CompactBinaryWriter<OutputBuffer>(output);

Transcode<DocumentKey>.FromTo(reader, writer);           
```

**Comment by sapek** [comment]
`Transcoder` changes the format (aka protocol), not the schema.

The schema specified to `Transcoder` describes schema of the payload. It is optional, e.g. transcoding between different tagged protocols doesn't require knowledge of the schema. However when transcoding from binary to text protocol or from untagged protocol you need to provide payload schema to supplement necessary metadata that is not in the payload.

**Comment by redknightlois** [comment]
Makes sense. However, being able to pass the metadata is a bit misleading. I stumbled upon it trying to abuse the transcoding mechanism to do what I cannot find how to do here. It wasnt that far fetched, if ```Transcode<T>``` would have worked in the same way as ```Deserialize<T>``` abiding to the same rules I could have built a non-standard write only protocol to deal with the extract payload without having to write  and maintain this insanely difficult expressions code. 

The expressions code is way outside of my domain of expertise, it's being a week trying to figure out how to write the proper expression. I am in the nearby of calling it (Bond) regardless of the performance improvements I could get. Mainly because I could not expect other devs to maintain the solution unless it is fool-proof.

**Comment by sapek** [comment]
Bond transforms are definitely an advanced usage scenario. As is often the case with meta-programming techniques, they allow you to achieve excellent performance but the implementation is more difficult.

If you prefer to stay away from custom expressions you have other options. You could filter the fields in a custom protocol writer (field's metadata is passed to `WriteFieldBegin` method). Or you could deserialize an instance of the key view and then serialize that.

There definitely could be a generic API that takes two schemas and projects payload from source schema into target schema. But a scenario where this would be useful has never come up before (as you've noted, yours is a somewhat unorthodox use case).
